### PR TITLE
Implement apply/cancel pattern

### LINKED ISSUE #2376: Implement Apply/Cancel pattern
Preferences need to be applied in one go in order to reduce the amount of time spent waiting for the VM to restart.

## Acceptance Criteria

- [x] The Apply and Cancel buttons are disabled if no settings have been changed
- [x] The Apply and Cancel buttons are enabled if one or more settings have been changed
- [x] Clicking the Apply button will commit all changed settings and begin VM restart if required
- [x] Clicking the Cancel button will discard all changed settings
- [x] Clicking the Cancel button will set all fields to the reflect the current settings object
- [x] Navigating or closing the window will warn that the changes won't be made, which would leave the earlier prompts to be handled
- [ ] e2e tests and unit tests are required to assert the behavior as described above

## Considerations

This means that we need an original copy of our preferences that will only be mutated when an explicit action takes place via `rdctl` or clicking the Apply button. This can be the settings object that we use today. We need a visual indicator that something changed while the UI is open. 

We need a mutable copy of our settings that we can use to track changes in the UI. This copy will be used to to compare changes with the original settings object before changes are applied. It will also be used to persist changes as a user navigates between tabs & can warn before navigating without applying changes.

- We need to track changes for enabling/disabling Apply/Cancel buttons
- We need to revert back to original values on Cancel click
- We need to apply new settings on Apply click. We would then close the current window and provide a status of the changes being applied.

### PR DESCRIPTION
**Pull request by rak-phillip**
This implements the apply/cancel pattern by making use of a vuex store to track preferences from a central location. With this pattern in place, components are responsible solely for rendering preference data and emitting events when said data needs to be updated. 

We identified issue #2509, which will need to be resolved before merging this PR.

closes #2376

### CODE REVIEW

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/pages/Preferences.vue [hunk 1]
```diff
@@ -16,22 +18,60 @@ export default Vue.extend({
...
+  computed: { ...mapGetters('preferences', ['getPreferences', 'isPreferencesDirty']) },
+  methods:  {
     navChanged(tabName: string) {
       this.currentNavItem = tabName;
     },
     closePreferences() {
       ipcRenderer.send('preferences-close');
+    },
+    async applyPreferences() {
+      await this.$store.dispatch(
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/pages/Preferences.vue:49]: We could introduce a watcher that will invoke `applyPreferences` only when credentials has a value. I originally wanted to await getting credentials, but there wasn't really a good way that I could find to do so with ipc/events..

File: src/pages/Preferences.vue [hunk 2]
```diff
@@ -16,22 +18,60 @@ export default Vue.extend({
...
+        'preferences/commitPreferences',
+        {
+          port:     this.credentials.port,
+          user:     this.credentials.user,
+          password: this.credentials.password
+        }
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/pages/Preferences.vue:55]: This is true. In hindsight, it's all so obvious 😅

File: src/window/preferences.ts [hunk 3]
```diff
@@ -30,5 +32,34 @@ export function openPreferences(parent: Electron.BrowserWindow) {
...
+    const result = dialog.showMessageBoxSync(
+      window,
+      {
+        title:   'Rancher Desktop - Close Preferences',
+        type:    'warning',
+        message: 'Close preferences without applying?',
+        detail:  'There are preferences with changes that have not been applied. All unsaved preferences will be lost.',
+        buttons: [
+          'Close preferences',
+          'Cancel'
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/window/preferences.ts:51]: I think `Discard changes` fits better here, thanks!

File: src/components/Preferences/ApplicationBehavior.vue [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
+
 import Checkbox from '@/components/form/Checkbox.vue';
 import RdFieldset from '@/components/form/RdFieldset.vue';
 
 export default Vue.extend({
   name:       'preferences-application-behavior',
   components: { Checkbox, RdFieldset },
+  props:      {
+    preferences: {
+      type:     Object,
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/components/Preferences/ApplicationBehavior.vue:12]: It's very possible 😁

File: src/components/Preferences/ApplicationBehavior.vue [hunk 5]
```diff
@@ -20,10 +26,18 @@ export default Vue.extend({
       statistics:       false,
     };
   },
-  // TODO: Move mapgetters up to page that will be managing the modal and replace
-  // usage with props
-  computed: { ...mapGetters('applicationSettings', ['sudoAllowed']) },
+  computed: {
+    isSudoAllowed(): boolean {
+      return !this.preferences?.kubernetes?.suppressSudo || false;
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/components/Preferences/ApplicationBehavior.vue:31]: Agreed, thanks!

File: src/components/Preferences/ApplicationBehavior.vue [hunk 6]
```diff
@@ -20,10 +26,18 @@ export default Vue.extend({
...
+    },
+    canAutoUpdate(): boolean {
+      return this.preferences?.updater || false;
+    }
+  },
   methods:  {
+    onChange(property: string, value: string | number | boolean) {
+      this.$store.dispatch('preferences/updatePreferencesData', { property, value });
```
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/components/Preferences/ApplicationBehavior.vue:39]: Yeah, this definitely could be improved upon. As it stands, this is quite fragile and breaks without much warning if properties in Settings are renamed..

File: src/pages/Preferences.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/pages/Preferences.vue:49]: After taking a closer look, I don't think the watcher makes sense..

We only assign credentials and attempt `fetchPreferences` when we receive the `api-credentials` event. We use the same credentials for any subsequent calls.

I think what would make more sense is for us to make sure that we have an error state for when we're unable to fetch credentials or preferences and supply a mechanism to retry in the event of a failure.

[the remark this answers is not in the document]
Comment (rak-phillip) [reply src/pages/Preferences.vue:49]: Created #2522 to track this work

**Review by mook-as — CHANGES_REQUESTED** [review]
Looks good generally, we can do things in follow ups.

Only issue is that the Kubernetes port number turns into a string in the Vuex state (so that it looks like it has changes to apply if you change the port number and back).

**Review by mook-as — APPROVED** [review]
LGTM; I think I can get the port to be `null`, but that needs backend help to reject invalid settings.

**Pull request merged**